server/models/RecurringExpense.ts

```diff
@@ -0,0 +1,219 @@
+      include: [
+        { model: models.Collective, as: 'collective' },
+        { model: models.ExpenseAttachedFile, as: 'attachedFiles' },
+        { model: models.ExpenseItem, as: 'items' },
+      ],
+    });
+    if (!expense) {
+      throw new Error(`Could not find previous expense for RecurringExpense #${this.id}`);
+    }
+
+    const draftKey = process.env.OC_ENV === 'e2e' || process.env.OC_ENV === 'ci' ? 'draft-key' : uuid();
+    const expenseFields = [
+      'description',
+      'longDescription',
+      'tags',
+      'type',
+      'privateMessage',
+      'invoiceInfo',
+      'PayoutMethodId',
+      'RecurringExpenseId',
+      'UserId',
+      'currency',
+    ];
+    const incurredAt = new Date();
+
+    const draft = {
+      ...pick(expense, expenseFields),
+      FromCollectiveId: this.FromCollectiveId,
+      CollectiveId: this.CollectiveId,
+      lastEditedById: expense.UserId,
+      incurredAt,
+      amount: expense.items?.reduce((total, item) => total + item.amount, 0) || expense.amount || 1,
+      data: {
+        items: expense.items?.map(item => ({ ...pick(item, ['amount', 'description', 'url']), incurredAt })),
+        attachedFiles: expense.attachedFiles?.map(file => pick(file, ['url'])),
+        payee: { id: expense.FromCollectiveId },
+        invitedByCollectiveId: this.FromCollectiveId,
+        draftKey,
+        payeeLocation: expense.payeeLocation,
+      },
+      status: expenseStatus.DRAFT,
+    };
+
+    const draftedExpense = await models.Expense.create(draft);
+    await this.update({ lastDraftedAt: incurredAt });
+
+    const inviteUrl = `${config.host.website}/${expense.collective.slug}/expenses/${draftedExpense.id}?key=${draft.data.draftKey}`;
+    await draftedExpense
+      .createActivity(
+        activities.COLLECTIVE_EXPENSE_RECURRING_DRAFTED,
+        { id: expense.UserId },
+        { ...draftedExpense.data, inviteUrl, description: draftedExpense.description },
+      )
+      .catch(e =>
+        console.error('An error happened when creating the COLLECTIVE_EXPENSE_RECURRING_DRAFTED activity', e),
+      );
+
+    return draftedExpense;
+  }
+
+  static async createFromExpense(
+    expense: typeof models.Expense,
+    interval: RecurringExpenseIntervals,
+    endAt?: string | Date,
+  ) {
+    if (typeof endAt === 'string') {
+      endAt = moment(endAt).toDate();
+    }
+
+    const recurringExpense = await this.create({
+      CollectiveId: expense.CollectiveId,
+      FromCollectiveId: expense.FromCollectiveId,
+      lastDraftedAt: new Date(),
+      interval,
+      endAt,
+    });
+    await expense.update({ RecurringExpenseId: recurringExpense.id });
+    return recurringExpense;
+  }
+
+  static async getRecurringExpensesDue() {
+    const dateWhere = Object.values(RecurringExpenseIntervals).map(interval => ({
+      lastDraftedAt: { [Op.lt]: moment().subtract(1, interval).endOf(interval).toDate() },
+      interval,
+    }));
```

Comment:
Nice way to build the conditions!

I wonder if the formula is correct though. Let's say that I have a monthly recurring expense processed on the 15th of each month. Today is the 1st of March, and the last time it was processed was the 15th of February. So the expected date for the recurring expense to tick is the 15th of march.

We're gonna look for:

```
lastDraftedAt: { LowerThan: NOW().subtract(1, MONTH).END_OF(MONTH) }
// => lastDraftedAt: { LowerThan: '1st March'.subtract(1, MONTH).END_OF(MONTH) }
// => lastDraftedAt: { LowerThan: '30th February' }
// => shouldn't that be?
// lastDraftedAt: { LowerThan: '15th March' }
```

Reply:
True. The END_OF interval should be consistent with the interval in which we run the cron job, it should always be "DAY".